### PR TITLE
Added timezones

### LINKED ISSUE #596: Select time zone for date module
It'd be extremely handy to be able to select which time zone a date module displays. Even just "utc = true" would be useful.

### LINKED ISSUE #1293: [module/date] Zone and UTC feature support request
Hello,

That's why I request to:

- Each user may want to see their own time zone. (Like I would like to see the time of Turkey and my brother may want to see Germany time.)
- We don't have to change the `zoneinfo` information from `/usr/share/` specially
- What time zone do we want to see ? LocalTime, UTC, Network Synched, etc...
- TimeZone offset support? UTC+1, UTC+2 like that.

Edit: Oh I forgot that trick, each user can have their own time zone. :)

Currently only localtime support is available with the [date] architecture. When we do `timedatectl`, there are few outputs:

```
Local time: Wed 2018-06-20 14:02:29 +03
Universal time: Wed 2018-06-20 11:02:29 UTC
RTC time: Wed 2018-06-20 14:02:29
Time zone: Europe/Istanbul (+03, +0300)
System clock synchronized: no
```

But Polybar uses localtime everytime as long as I do not sync with localtime to network time using timezones from mentioned here https://wiki.archlinux.org/index.php/Time#Time_zone

I don't think that's a rational solution for Polybar. The time module should be further improved. The module I designed in my mind is:

```
[module/date]
type = internal/date

; New date features
timezone = Europe/Istanbul ; from /usr/share/zoneinfo
timezoneoffset = +3, +2, 0, bla bla...
datetype = UTC, RTC, Local, Universal etc...
autosync = true ; Possibility of system time not synced


; Seconds to sleep between updates
interval = 1.0

; See "http://en.cppreference.com/w/cpp/io/manip/put_time" for details on how to format the date string
; NOTE: if you want to use syntax tags here you need to use %%{...}
date = %Y-%m-%d%

; Optional time format
time = %H:%M

; if `date-alt` or `time-alt` is defined, clicking
; the module will toggle between formats
date-alt = %A, %d %B %Y
time-alt = %H:%M:%S
```

All comments are welcomed. Thanks in advance! :)

### PR DESCRIPTION
**Pull request by thomastay**
Added the ability to display two clocks on the date module and the ability to set the timezone on those two clocks. This solves #596 and also addresses the problem of #788 in that the fix does not use any external modules, instead relying on setting the TZ environment variable. 

Currently the options for customization are such that users can set the setting **time-TZ** and **time-2-TZ** on their polybar config file to set the timezones of two separate clocks. The options date-2 and time-2 can be added to the label to give the ability to display a dual clock. date-2 is set using the time-2-TZ variable and is not independent of time-2. 

Here's an example using my polybar config file:
```
type = internal/date
interval = 1

date =  %a, %b %d
date-alt =  %Y-%m-%d 

time = "%H:%M"
time-TZ = "Asia/Singapore"
time-2-TZ = "America/Detroit"
time-alt = "%H:%M:%S"

label = %date%   %time% | %date-2%   %time-2%
```
My config

Minor bugs: currently, setting the option time-2-TZ but _not_ time-TZ results in both clocks having the secondary time zone. I don't know how to fix this bug at the moment.

Improvements: I intend to work on a version of this module that allows users to have different timezone when they click on the module, instead of just changing the format.

Fixes #596 
Fixes #1293

### CODE REVIEW

**Comment by patrick96** [comment]
What is the benefit of dual clocks in the same module over having two separate date modules?

**Comment by thomastay** [comment]
@patrick96 I suppose you're right. I originally designed the dual clock system with the intention of having a click-to-change timezone module, that's why I implemented the option for dual clock. However that proved difficult without completely rewriting the module so that's why I added the option to display two clocks at once.

**Comment by thomastay** [comment]
hmm I tried having two modules with one timezone each, but I ran into a problem that having two date modules means that the click handler only applies to first of the modules defined. This means that on clicking either of the modules, only the first module defined changes to 

Is this a known bug? I've looked through the code and I'm not sure why the

**Comment by patrick96** [comment]
> Is this a known bug?

Yes, this is the same issue as #1172. If you have multiple modules of the same kind that have internal click events, they are all caught and processed by the first module and don't reach the others.

**Comment by thomastay** [comment]
Ok! Have done so.

**Comment by thomastay** [comment]
@patrick96 are there any more changes that need to be made to the PR before it can be accepted?

**Comment by patrick96** [comment]
Probably. I haven't had the time to properly review this.

File: src/modules/date.cpp [hunk 1]
```diff
@@ -41,6 +42,14 @@ namespace modules {
   bool date_module::update() {
     auto time = std::time(nullptr);
 
+    //Timezone format business
+    const char* timezone_format_original = getenv("TZ");
+
+    string timezone_format = m_time_timezone;
+    if (timezone_format != ""){
```
Comment (patrick96) [inline src/modules/date.cpp]: Just a style thing: Please use a space after the if parenthesis

**Comment by NBonaparte** [comment]
We might need to use a mutex to prevent race conditions between multiple modules.

**Comment by thomastay** [comment]
@patrick96 ok i will look into it. Using a mutex like @NBonaparte mentioned is probably the best solution.

**Comment by thomastay** [comment]
@patrick96 I've thought about this and here's a solution I've come up with to prevent the race condition (it is indeed a race condition, I checked). Wanted to hear your input on this as I'm not an expert on multithreading.

Have a central mutex held by the bar's _controller_. When a date module wants to access / set an environment variable, it needs to use the mutex held by the bar's controller. This ensures that only one module is accessing the "TZ" environment variable at once. 

Right now the modules cannot directly call this central mutex since they don't inherit a pointer to the controller. I was thinking of modifying the start function to include a pointer to the controller that started it, but that might be unwieldy. Any ideas / opinions?

**Comment by patrick96** [comment]
Yeah right now we don't really have central place to store such things where everyone has access to. I think storing the mutex in `bar_settings` in `types.hpp` for now should suffice, every module already has access to that. At some point we should probably rethink how things are coupled in general and maybe use a service locator for such things.

**Comment by patrick96** [comment]
@thomastay are you still working on this?

**Comment by thomastay** [comment]
sorry, I haven't had time to work on it. Not working on it for now.

**Comment by patrick96** [comment]
No worries. Maybe someone else will pick this up in the future :)

**Pull request closed without merging**